dotnet/src/Skills/Skills.Grpc/Extensions/KernelGrpcExtensions.cs

```diff
@@ -0,0 +1,190 @@
+        {
+            try
+            {
+                kernel.Log.LogTrace("Registering gRPC function {0}.{1}", skillName, operation.Name);
+                var function = kernel.RegisterGrpcFunction(skillName, operation);
+                skill[function.Name] = function;
+            }
+            catch (Exception ex) when (!ex.IsCriticalException())
+            {
+                //Logging the exception and keep registering other gRPC functions
+                kernel.Log.LogWarning(ex, "Something went wrong while rendering the gRPC function. Function: {0}.{1}. Error: {2}",
+                    skillName, operation.Name, ex.Message);
+            }
+        }
+
+        return skill;
+    }
+
+    #region private
+
+    /// <summary>
+    /// Registers SKFunction for a gRPC operation.
+    /// </summary>
+    /// <param name="kernel">Semantic Kernel instance.</param>
+    /// <param name="skillName">Skill name.</param>
+    /// <param name="operation">The gRPC operation.</param>
+    /// <returns>An instance of <see cref="SKFunction"/> class.</returns>
+    private static ISKFunction RegisterGrpcFunction(
+        this IKernel kernel,
+        string skillName,
+        GrpcOperation operation)
+    {
+        var operationParameters = operation.GetParameters();
+
+        async Task<SKContext> ExecuteAsync(SKContext context)
+        {
+            try
+            {
+                var arguments = new Dictionary<string, string>();
+
+                //Extract function arguments from context
+                foreach (var parameter in operationParameters)
+                {
+                    //A try to resolve argument parameter name.
+                    if (context.Variables.Get(parameter.Name, out var value))
+                    {
+                        arguments.Add(parameter.Name, value);
+                        continue;
+                    }
+
+                    throw new KeyNotFoundException($"No variable found in context to use as an argument for the '{parameter.Name}' parameter of the '{skillName}.{operation.Name}' gRPC function.");
+                }
+
+                var runner = new GrpcOperationRunner(new HttpClient());
+
+                //SKFunction should be extended to pass cancellation token for delegateFunction calls.
+                var result = await runner.RunAsync(operation, arguments, CancellationToken.None).ConfigureAwait(false);
+
+                if (result != null)
+                {
+                    context.Variables.Update(result.ToString());
+                }
+            }
+            catch (Exception ex) when (!ex.IsCriticalException())
+            {
+                kernel.Log.LogWarning(ex, "Something went wrong while rendering the gRPC function. Function: {0}.{1}. Error: {2}", skillName, operation.Name,
+                    ex.Message);
+                context.Fail(ex.Message, ex);
+            }
+
+            return context;
+        }
+
+        // TODO: to be fixed later
```

Comment:
can `using` here resolve this issue?

Reply:
Unfortunately, it won't. If we put using here the instance of the SKFunction will be disposed as soon as execution control leaves this registration method way before the function got executed. It's not what we want.

At the moment, we don't need to do anything here because the instance will be registered in Kernel's skillsCollection which we plan to make disposable so that all functions registered in the collection will be disposed automatically when Kernel.Dispose method is called.

An alternative approach would be to move responsibility of managed resources creation out of SK SDK to a hosting application. As result, SK SDK would not need to dispose them because it did not create them and does not own them. That would simplify SK SDK code and would give more control over the resources lifetime to the hosting app which could choose to declare them as singletons and retain them for the entire app lifetime or have them as transient ones - create an instance per request.